source/allocator_sba.c

```diff
@@ -0,0 +1,363 @@
+        s_aligned_free(page);
+        return;
+    }
+
+    aws_array_list_push_back(&bin->chunks, &addr);
+}
+
+/* No lock required for this function, it's all read-only access to constant data */
+static struct free_bin *s_sba_find_bin(struct small_block_allocator *sba, size_t size) {
+    AWS_PRECONDITION(size <= s_max_bin_size);
+    for (unsigned idx = 0; idx < AWS_SBA_BIN_COUNT; ++idx) {
+        struct free_bin *bin = &sba->bins[idx];
+        if (bin->size >= size) {
+            return bin;
+        }
+    }
+    /* should never get here */
+    AWS_POSTCONDITION(false);
+    return NULL;
+}
+
+static void *s_sba_alloc(struct small_block_allocator *sba, size_t size) {
+    if (size <= s_max_bin_size) {
+        struct free_bin *bin = s_sba_find_bin(sba, size);
+        AWS_FATAL_ASSERT(bin);
+        /* BEGIN CRITICAL SECTION */
+        aws_mutex_lock(&bin->mutex);
+        void *mem = s_sba_alloc_from_bin(bin);
+        aws_mutex_unlock(&bin->mutex);
+        /* END CRITICAL SECTION */
+        return mem;
+    }
+    return aws_mem_acquire(sba->allocator, size);
+}
+
+static void s_sba_free(struct small_block_allocator *sba, void *addr) {
+    if (!addr) {
+        return;
+    }
+
+    struct page_header *page = (struct page_header *)s_page_base(addr);
+    if (page->tag == AWS_SBA_TAG_VALUE) {
+        struct free_bin *bin = page->bin;
+        /* BEGIN CRITICAL SECTION */
+        aws_mutex_lock(&bin->mutex);
+        s_sba_free_to_bin(bin, addr);
+        aws_mutex_unlock(&bin->mutex);
+        /* END CRITICAL SECTION */
+        return;
+    }
+    /* large alloc, give back to underlying allocator */
+    aws_mem_release(sba->allocator, addr);
+}
+
+static void *s_sba_mem_acquire(struct aws_allocator *allocator, size_t size) {
+    struct small_block_allocator *sba = allocator->impl;
+    return s_sba_alloc(sba, size);
+}
+
+static void s_sba_mem_release(struct aws_allocator *allocator, void *ptr) {
+    struct small_block_allocator *sba = allocator->impl;
+    s_sba_free(sba, ptr);
+}
+
+static void *s_sba_mem_realloc(struct aws_allocator *allocator, void *old_ptr, size_t old_size, size_t new_size) {
+    struct small_block_allocator *sba = allocator->impl;
+    /* If both allocations come from the parent, let the parent do it */
+    if (old_size > s_max_bin_size && new_size > s_max_bin_size) {
+        void *new_mem = old_ptr;
+        if (aws_mem_realloc(sba->allocator, &new_mem, old_size, new_size)) {
+            return NULL;
+        }
+        return new_mem;
+    }
+
+    /* Either both small allocs, or one small, one large */
+    void *new_mem = s_sba_alloc(sba, new_size);
+    memcpy(new_mem, old_ptr, aws_min_size(old_size, new_size));
+    s_sba_free(sba, old_ptr);
+    return new_mem;
+}
+
+static void *s_sba_mem_calloc(struct aws_allocator *allocator, size_t num, size_t size) {
+    struct small_block_allocator *sba = allocator->impl;
+    void *mem = s_sba_alloc(sba, size * num);
+    memset(mem, 0, size * num);
```

Comment:
[optimization] we could have the `page_cursor` be zeroed to start
then you'd only need to memset a previously freed chunk

Reply:
That would also cache the page. I should profile it both ways. It may have a negative effect.

Reply:
It didn't seem to make a measurable difference, but I'm only doing micro-benchmarks. Once we have a load to test it on, I can try it again.

Reply:
good call, probably not worth caching the page